domain/user/service/service.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/juju/juju/core/user"
+	usererrors "github.com/juju/juju/domain/user/errors"
+)
+
+// State describes retrieval and persistence methods for upgrade info.
+type State interface {
+	// AddUser will add a new user to the database. If the user already exists
+	// an error that satisfies usererrors.AlreadyExists will be returned.
+	AddUser(context.Context, string, string, string, string) error
+
+	// GetUser will retrieve the user specified by name from the database where
+	// the user is active and has not been deleted. If the user does not exist
+	// or is deleted an error that satisfies usererrors.NotFound will be
+	// returned.
+	GetUser(context.Context, string) (user.User, error)
+	// RemoveUser marks the user as deleted. This obviates the ability of a user
```

Comment:
@anvial  "marks the user as removed", I think we should also expand the scope of what this func does to include the fact that it will "delete" all authentication information associated with the user.